cache/disk/disk.go

```diff
@@ -85,13 +85,55 @@ func New(dir string, maxSizeBytes int64, proxy cache.CacheProxy) *DiskCache {
 	}
 
 	// The eviction callback deletes the file from disk.
+	// This function is only called while the lock is held
+	// by the current goroutine.
 	onEvict := func(key Key, value SizedItem) {
-		// Only remove committed items (as temporary files have a different filename)
+		
+		f := filepath.Join(dir, key.(string))
+
 		if value.(*lruItem).committed {
-			err := os.Remove(filepath.Join(dir, key.(string)))
+			// Common case. Just remove the cache file and we're done.
+			err := os.Remove(f)
 			if err != nil {
-				log.Println(err)
+				log.Printf("ERROR: failed to remove evicted cache file: %s", f)
 			}
+
+			return
+		}
+
+		// There is an ongoing upload for the evicted item. The temp
+		// file may or may not exist at this point.
+		//
+		// We should either be able to remove both the temp file and
+		// the regular cache file, or to remove just the regular cache
+		// file. The temp file is renamed/moved to the regular cache
+		// file without holding the lock, so we must try removing the
+		// temp file first.
+
+		// Note: if you hit this case, then your cache size might be
+		// too small (blobs are moved to the most-recently used end
+		// of the index when the upload begins, and these items are
+		// still uploading when they reach the least-recently used
+		// end of the index).
+
+		tf := f + ".tmp"
+		var fErr, tfErr error
+		removedCount := 0
+
+		tfErr = os.Remove(tf)
```

Comment:
Two thoughts:
1) Instead of only deleting the file shouldn't it rather also cancel the upload?
2) I believe one Windows one can't delete a file that's open. That being said I don't know if anyone runs bazel-remote on Windows but it should sure be taken into consideration.

Reply:
> 1. Instead of only deleting the file shouldn't it rather also cancel the upload?

Maybe, but that would require a bit of refactoring, eg storing a context in the lruItem.

Or maybe the item should be given some bias and periodically moved to the MRU end of the index while it's being uploaded (with some deadline)?

But IMO this is mostly a sign that the cache size is too small for this use case.

> 2. I believe one Windows one can't delete a file that's open. That being said I don't know if anyone runs bazel-remote on Windows but it should sure be taken into consideration.

Right. I think this would also be a problem in the regular eviction case on windows- there can be ongoing downloads of an item that is evicted, then we can't remove the file. To handle windows well would probably require a fancy file management layer, or some ugly retry loops.